### PR TITLE
Pagination: Pagination feature/compute index

### PR DESCRIPTION
**Pull request by staheri14**
This PR addresses [this issue](https://github.com/vacp2p/specs/issues/215).

### CODE REVIEW

File: waku/node/v2/waku_types.nim
Comment (oskarth) [inline waku/node/v2/waku_types.nim:178]:
```suggestion
proc computeIndex* (w: WakuMessage): Index =
```

File: waku/node/v2/waku_types.nim
Comment (oskarth) [inline waku/node/v2/waku_types.nim:180]:
```suggestion
  let payloadStr=w.payload.join("") # converts payload to string
```

File: waku/node/v2/waku_types.nim
Comment (oskarth) [inline waku/node/v2/waku_types.nim:185]:
```suggestion
proc computeIndex*(w: WakuMessage): Index =
  ## Takes a WakuMessage and returns its index
  let payloadStr = w.payload.join("") # converts payload to string
  var data: string
  data.add(w.contentTopic)
  data.add(payloadStr)
  result.digest = $secureHash(data) # computes sha1 of data
  result.timeStamp = $getTime().utc # gets the timestamp in utc with seconds precision
```

Comment (oskarth) [reply waku/node/v2/waku_types.nim:185]: you can also use `&` for string concatenation, so it could probably be written something like (untested, assuming $ is implemented on payload):

let data = msg.contentTopic & $msg.payload

Comment (oskarth) [reply waku/node/v2/waku_types.nim:185]: Few more comments:

- I'd make it clear that the timestamp is for receiver node, so `receivedTime` or something like this.

- I believe contentTopic is optional, so might want to make sure this is handled.

- Perhaps a test here would be useful to make sure it works as intended?

**Review by oskarth — COMMENTED** [review]
See above

File: waku/node/v2/waku_types.nim
Comment (kdeme) [inline waku/node/v2/waku_types.nim:41]: I think string is not needed here and `uint32` could be used as this is probably going to be a unix timestamp?

File: waku/node/v2/waku_types.nim
Comment (kdeme) [inline waku/node/v2/waku_types.nim:185]: What about `epochTime`, wouldn't that be easier to work with?

Comment (staheri14) [reply waku/node/v2/waku_types.nim]: I can also use `epochTime`, it gives a `float` value, however, the documentation of `epochTime` proc recommends  `getTime` instead of `epochTime`:
> "``getTime`` should generally be preferred over this proc."
I am not sure why.

But storage-wise, float is better than string.

File: waku/node/v2/waku_types.nim
Comment (staheri14) [reply waku/node/v2/waku_types.nim:41]: I will change the timestamp to unix timestamp (which is of type `float`)

File: waku/node/v2/waku_types.nim
Comment (staheri14) [reply waku/node/v2/waku_types.nim]: Thanks! I will use nimcrypto instead, could you also please give me a resource for the insecurity of stdlib hash implementation, I would like to read more about it.

File: waku/node/v2/waku_types.nim
Comment (staheri14) [reply waku/node/v2/waku_types.nim]: > Perhaps a test here would be useful to make sure it works as intended?

I was going to implement the tests in the next issue, but per your comment, I'll include it in this PR.

Comment (staheri14) [reply waku/node/v2/waku_types.nim]: > I'd make it clear that the timestamp is for receiver node, so receivedTime or something like this.
> 
> I believe contentTopic is optional, so might want to make sure this is handled.

Thanks, addressed!

**Comment by oskarth** [comment]
See https://github.com/status-im/nim-waku/pull/224#issuecomment-708884595 re branches etc. This PR also needs to be rebased on master I believe.

@staheri14 ping here / re-request reviews once it is ready!

File: waku/node/v2/waku_types.nim
Comment (kdeme) [reply waku/node/v2/waku_types.nim]: I'm honestly not really sure if it is insecure, or how insecure it is. We use nimcrypto hashes in our projects and that code has been audited for our nimbus-eth2 project.  So I put a lot more thrust in that codebase. I don't think the stdlib code has ever been audited, but I might be mistaken.

Additionally, sha1 should not be used any more for secure hashing in new projects. Stdlib doesn't have any other hashing implementations (aside from MD5, heh lets not go there). 
Also, a smaller item, the APIs from the stdlib are different for the different hash implementations, which is just weird.

File: waku/node/v2/waku_types.nim
Comment (kdeme) [inline waku/node/v2/waku_types.nim:42]: I would hold this value here as `MDigest[256]`, and send it over the wire as byte array.

File: waku/node/v2/waku_types.nim
Comment (kdeme) [inline waku/node/v2/waku_types.nim:187]: When stored as MDigest, you can just use something like:
```nim 
var ctx: keccak256
ctx.init()
ctx.update(msg.contentTopic.toBytes()) # Use the one from nim-stew byteutils
ctx.update(msg.payload)
let digest = ctx.finish()
```

File: waku/node/v2/waku_types.nim
Comment (staheri14) [reply waku/node/v2/waku_types.nim:42]: Sure, done

**Comment by staheri14** [comment]
> See [#224 (comment)](https://github.com/status-im/nim-waku/pull/224#issuecomment-708884595) re branches etc. This PR also needs to be rebased on master I believe.
> 
> @staheri14 ping here / re-request reviews once it is ready!

@oskarth, I rebased the current PR to master, it is ready for merge with master.

**Review by oskarth — APPROVED** [review]
LGTM!

Once @decanus approves, feel free to merged to master at your own leisure @staheri14. The only thing to make sure it is rebased on latest master and that CI (at least Ubuntu) passes.

If you are unable to merge for some reason (access control or w/e), please ping @decanus or me and we can merge it.

File: tests/v2/test_waku_pagination.nim
Comment (decanus) [inline tests/v2/test_waku_pagination.nim:7]: @staheri14 we use 2 spaces for indentation.

**Review by decanus — CHANGES_REQUESTED** [review]

File: tests/v2/test_waku_pagination.nim
Comment (decanus) [inline tests/v2/test_waku_pagination.nim:25]: @staheri14 this all should be 2 space indentation rather than 4. The mix won't work.

**Review by decanus — CHANGES_REQUESTED** [review]

File: tests/v2/test_waku_pagination.nim
Comment (staheri14) [reply tests/v2/test_waku_pagination.nim]: Thanks! Addressed the indentation issues.

File: waku/node/v2/waku_types.nim
Comment (kdeme) [inline waku/node/v2/waku_types.nim:7]:
```suggestion
  std/[tables, times],
```
Not necessary and more of a nitpick but something we've recently started doing in the Nimbus team is put `std/` in front of all stdlib imports, that way it is immediately visible which imports are from the stdlib. 
Also, I guess sha1 can be dropped here?

**Comment by oskarth** [comment]
I think you missed this review yesterday @decanus? Looks like it was just formatting that was addressed, going ahead and merging.

**Pull request merged**
